File: cpp/46_goldbachs_other_conjecture.cpp
#include <iostream>

bool is_prime(int n){
    for (int i = 2; i * i <= n; i++){
        if (n % i == 0) return false;
    }
    return true;
}

bool is_goldbach_sum(int n){
    // a composite number n is a Goldbach sum
    // if it can be written as n = p + 2*s*s,
    // the sum of a prime p and twice a square (s^2)
    for (int s = 1; 2 * s * s < n; s++){
        int p = n - 2 * s * s;
        if (is_prime(p)) return true;
    }
    return false;
}

int goldbachs_other_conjecture(){
    int res = 0, n = 35;
    while (res == 0){
        // check if current composite n is a Goldbach sum
        if (!is_prime(n) and !is_goldbach_sum(n)) res = n;
        n += 2;
    }
    return res;
}

int main(){
    std::cout << goldbachs_other_conjecture() << "\n";
}

File: java/_46_GoldbachsOtherConjecture.java
import java.util.*;

public class _46_GoldbachsOtherConjecture{
    
    public static boolean isPrime(int n){
        for (int i = 2; i*i < n; i++){
            if (n % i == 0)
                return false;
        }
        return true;
    }
    
    public static void main(String[] args){
        
        // If n = p + 2*s^2, then p = n - 2*s^2.
        // Starting at n = 35, apply the algorithm to every odd composite number.
        // Iterate through all values of i such that i*i = s and 2*s < n,
        // then check if the difference is prime.
        // If all values of i have been checked and no corresponding sum has been found,
        // n is the smallest number that contradicts Goldbach's conjecture.
        
        int smallest = 0;
        // 33 can be written as a sum of a prime and twice a square. (33 = 31 + 2 * 1^2)
        int n = 35; 
        while (smallest == 0){
            
            if (isPrime(n)){
                // System.out.println(n + " is prime");
                n += 2;
                continue;
            }
            
            boolean foundSum = false;
            for (int i = 1; 2*i*i < n; i++){
                int s = i*i;
                if (isPrime(n - 2*s)){
                    // System.out.printf("%d = %d + 2*%d\n", n, n - 2*s, s);
                    foundSum = true;
                    break;
                }
            }
            
            if (!foundSum)
                smallest = n;
            n += 2;
        }
        
        System.out.println("Smallest odd composite that cannot be written as the sum of a prime and twice a square: " + smallest); // 5777
    }
}